### PR TITLE
Page comparison failures include wildcards #4170

### LINKED ISSUE #4170: Page comparison failures include wildcards
Exclude wildcards from comparisons for better readability

### PR DESCRIPTION
**Pull request by bamboo3250**
Fix #4170
Basically, when a difference is detected, we change real values, which are supposed to be wildcards, back to wildcards then do the comparison.

### CODE REVIEW

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 1]
```diff
@@ -742,8 +742,21 @@ private boolean testAndRunGodMode(String filePath, String content) {
             return false;
         }
     }
-
-    private String processPageSourceForGodMode(String content) {
+    
+    public static String processPageSourceForFailureCase(String content) {
+        return processPageSourceForGodMode(content)
+                // jQuery local
+                .replace("&#x2f;js&#x2f;lib&#x2f;jquery.min.js", "{*}&#x2f;jquery.min.js")
```
Comment (LowWeiLin) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Can we make this more readable? Use a function to encode the url instead of writing the url in encoded form here.

Comment (bamboo3250) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Yes, we can use `Sanitizer.sanitizeForHtml` for that.

Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Why is this needed? I thought God mode already included this?

**Comment by bamboo3250** [comment]
@LowWeiLin Done. Please review it.

File: src/test/java/teammates/test/pageobjects/AppPage.java (hunk 1 above)
Comment (bamboo3250) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: What Godmode have is the "raw" version of those strings. When the web page source is processed, its content will be sanitized so Godmode cannot convert those strings back to `{*}`.

Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Goodness, that means God mode is broken. :scream:

Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: My fear is confirmed: GodModeTest fails. Will open an issue on this.

Comment (bamboo3250) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: I'm not sure why God mode is broken. Could you explain more about it?

Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: See #4202

**Comment by LowWeiLin** [comment]
i forgot there was a Godmode test. Do add/modify a test to include this case.

**Comment by wkurniawan07** [comment]
Keep this on hold first. The fix for #4202 may make the extra replaces in `processPageSourceForFailureCase` unnecessary.

**Comment by wkurniawan07** [comment]
See if you still need the extra replaces now.

**Comment by bamboo3250** [comment]
I merged the current master and those replaces are still needed since before being compared, the "expected" and "actual" strings are also sanitized in the function `convertToStandardHtmlRecursively`. But what you did is still correct and necessary anyway.

**Comment by wkurniawan07** [comment]
Oh, right. In that case, you should also add this one other case that's sanitized: line 785 of `AppPage.java`.

**Comment by bamboo3250** [comment]
You mean these 3 lines? 
```
 .replaceAll("([a-zA-Z0-9-_]){30,}%"
                        + "[\\w+-][\\w+!#$%&'*/=?^_`{}~-]*+(\\.[\\w+!#$%&'*/=?^_`{}~-]+)*+@([A-Za-z0-9-]+\\.)*[A-Za-z]+%"
                        + "[\\w+-][\\w+!#$%&'*/=?^_`{}~-]*+(\\.[\\w+!#$%&'*/=?^_`{}~-]+)*+@([A-Za-z0-9-]+\\.)*[A-Za-z]+", "{*}")
```

**Comment by wkurniawan07** [comment]
Lol no, okay I should've just copied the line here:
```
// today's date
.replace(TimeHelper.formatDate(now), "{*}")
```
This is because that function returns `DD/MM/YYYY`, the same problem.

**Comment by bamboo3250** [comment]
Okay, no problem :+1:

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 2]
```diff
@@ -742,8 +743,28 @@ private boolean testAndRunGodMode(String filePath, String content) {
...
     }
-
-    private String processPageSourceForGodMode(String content) {
+    
+    public static String processPageSourceForFailureCase(String content) {
+        Date now = new Date();
+        return processPageSourceForGodMode(content)
+                // jQuery local
+                .replace(Sanitizer.sanitizeForHtml("/js/lib/jquery-ui.min.js"), 
+                         Sanitizer.sanitizeForHtml("{*}/jquery-ui.min.js"))
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Lol, duplicate `jquery-ui.min`. The `jquery.min` is not changed.

Comment (bamboo3250) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: oh my brain must have been not working at that time, creating an issue now :sweat:

**Comment by wkurniawan07** [comment]
Actually, why do we even need to sanitize during comparison...? @bamboo3250 I found out that you're the one who added the sanitizing part in the `HtmlHelper`, but for what?

**Comment by bamboo3250** [comment]
@wkurniawan07 Okay, let's say we have:

`https://ajax.googleapis.com/ajax/libs/jqueryui/1.10.4/jquery-ui.min.js` in the actual web page source and
`{*}/jquery-ui.min.js` in the html output.

When we do `assertSameHtml()` or `assertSameHtmlPart()`, they have to go through `convertToStandardHtml()`, which will sanitize them to 
`https:&#x2f;&#x2f;ajax.googleapis.com&#x2f;ajax&#x2f;libs&#x2f;jqueryui&#x2f;1.10.4&#x2f;jquery-ui.min.js` and `{*}&#x2f;jquery-ui.min.js`.

If there is a difference after comparing 2 files, we want to ignore those 2 lines in the comparison so I have to make those two look similar by changing them back to {*}&#x2f;jquery-ui.min.js`.

The things like `.replace(Sanitizer.sanitizeForHtml("/js/lib/jquery-ui.min.js"), Sanitizer.sanitizeForHtml("{*}/jquery-ui.min.js"))` are for that purpose.

**Comment by wkurniawan07** [comment]
What I'm asking is why do we need to sanitize in `convertToStandardHtml()`, which is why we need to go through this additional `replace`s. If `convertToStandardHtml()` doesn't have sanitizing, all these `replace`s are unnecessary. And I'm of the belief that it _is_ unnecessary.

**Comment by bamboo3250** [comment]
I think you have a point. I should have added the sanitization in jsp or PageData instead of HtmlHelper. Even for now, after I tried removing those unnecessary sanitization, the test still passed so we totally can remove them. I'll create an issue to undo that.
